### DISCUSSION TITLE
Istio 1.22.x gives warnings on envoyfilters

### DISCUSSION THREAD

**Discussion by kkwekkeboom** [istio/istio#51414] [General]
After an upgrade to istio 1.22 we get warnings when adding envoyfilters.

My use case is to implement rate limiting using envoyfilters, which is a pretty common use case I suspect, see https://istio.io/latest/docs/tasks/policy-enforcement/rate-limit/. using `kubectl apply -f ...` I get warnings
`EnvoyFilter exposes internal implementation details that may change at any time. Prefer other APIs if possible, and exercise extreme caution, especially around upgrades.`

This is a new addition in istio 1.22 https://github.com/istio/istio/blob/915c07216931/pkg/config/validation/envoyfilter/envoyfilter.go#L42 and assumes most functionality is replaced already by istio V1 APIs, which is not the case https://istio.io/latest/blog/2024/v1-apis/#overview-of-istio-crds. 

Is this warning not a bit premature to add to istio?

**Comment by howardjohn** [comment]
No, its not. The warning does NOT assume functionality is already replaced by other APIs, it aims to warn you about the real risk of using Envoy Filter **even if there is no alternatives**.

The fact people _don't believe the warning_ (you are not alone, a few others mentioned similar) is a pretty good sign, IMO, that it was a good idea to add. There is a huge mismatch between the maintainer and user perception of the stability and risk of the API.

**Reply by howardjohn** [reply]
Also keep in mind its just a warning. You are free to ignore it

**Reply by kkwekkeboom** [reply]
Thanks for the quick response!

Since envoy is quite embedded in Istio any change in envoy has impact on Istio, especially features mentioned in the Istio docs. For me this warning signals I'm doing something wrong. And definitely I'm believing the warning and checked thoroughly my rate limiting still works.I'm aware the envoyfilter is in alpha3 stage.

Would there be an option to enable/disable this warning? Or is there a roadmap to make rate limiting part of the Istio V1 APIs?

**Reply by howardjohn** [reply]
Istio insulates users from Envoy changes via its APIs -- except with EnvoyFilter. And its not just envoy changes, we may change how Istio generates XDS, etc. We break envoyfilter all the time, probably 50% of releases have some breakage.

There isn't any option to disable the warning. there has been a lot of discussion on rate limitting but no concrete roadmaps

**Reply by kkwekkeboom** [reply]
If there is any proposal for rate limiting I would definitely upvote that. For now I will ignore the warning and test the envoyfilters used by any upgrade of istio. Thanks for the quick responses @howardjohn

**Comment by taitelman** [comment]
is it possible to write the name of the envoyFilter in question. its a crucial information for big helm charts.

**Reply by howardjohn** [reply]
The warning happens on literally _any_ EnvoyFilter, so it seems easy you don't really need the name?

**Comment by willismorse** [comment]
I fully appreciate the risks of using EnvoyFilter and i wish we could avoid it. But we do have a small number of custom Envoy filters that we needed to create.

Is there any other safer Istio alternative for integrating custom Envoy filters?

**Comment by armingerten** [comment]
> There is a huge mismatch between the maintainer and user perception of the stability and risk of the API.

Even the official documentation (https://istio.io/latest/docs/tasks/policy-enforcement/rate-limit/) suggests that EnvoyFilters are "the way to go" to configure rate limiting. As a user, I tend to rely on the official documentation and unless there is an explicit disclaimer (e.g. "this is an alpha feature" or "this is using an unstable API"), I perceive a feature to be stable.

> Also keep in mind its just a warning. You are free to ignore it

If the warning is supposed to be ignored, then it's not supposed to be a warning. As a user, I need to be able to fix the behavior that is causing the warning by e.g. altering my configuration. I'd be happy to configure rate limiting in any other way, but from what I understand, the EnvoyFilter is the suggested implementation to achieve this?

**Reply by ilrudie** [reply]
That's a good call out. I'll try to update the docs to make it more clear up front that you will be using an unstable and risky api to implement rate limits.

The warning is certainly not meant to be ignored. You are free to ignore, if you value what you are getting from EF enough and are willing to shoulder the risks yourself.
